# Basic Pandas Dataframe Manipulation

Pandas is widely regarded as one of Python's most used libraries for data science, and there are a variety of online course, youtube videos and blogs about its use.  This project and blog post is aimed at providing new Python users with some of the basics of manipulating the Pandas dataframe.  I will cover the following areas:

1. Examining Your Data
2. Sorting and Subsetting
3. Calculating Summary Statistics
4. Creating Pivot Tables
5. Using Explicit Indexes

But first, we'll import some data.  I am a competitive cyclocross racer and track a large amount of data about my training, performance and sleep habits over time.  This dataset contains daily data on these metrics, and contains both dimension and measure columns.

In [131]:
# Import Libraries

import pandas as pd
import numpy as np

# Get the datafile from my public GitHub repo
url = 'https://raw.githubusercontent.com/joelmsherman/explain_python-dataframe/master/HRV_Training_Sleep_clean.csv'

# Read the content and turn it into a Pandas Dataframe
AllData = pd.read_csv(url, error_bad_lines=False)


## 1. Examining Your Data

We can take a quick look at the dataframe with some simple commands:

df.head() returns the first few rows of the dataframe.<br>
df.info() provides info about the columns, including their index number, names, non-null counts and data types.<br> 
df.shape() provides a tuple that shows the total number of rows, and total number of columns.<br>
df.describe() computes summary statistics for the numeric (float) columns types.<br>

Let's just look at info()

In [132]:
print(AllData.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 213 entries, 0 to 212
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   date                   213 non-null    object 
 1   HR_lying               203 non-null    float64
 2   rMSSD_lying            203 non-null    float64
 3   HRV4T_Recovery_Points  203 non-null    float64
 4   training               213 non-null    object 
 5   physical_condition     65 non-null     float64
 6   trainingRPE            134 non-null    float64
 7    trainingTSS           124 non-null    float64
 8    sleep_quality         204 non-null    float64
 9   sleep_time             213 non-null    float64
 10  muscle_soreness        200 non-null    float64
 11  fatigue                200 non-null    float64
 12  traveling              213 non-null    object 
 13  sickness               213 non-null    object 
 14  alcohol                213 non-null    object 
 15   curre

We can see that the dataframe AllData has 213 rows, and 16 columns, 10 of which are numeric and 6 non-numeric.  Further, some columns contain many null values, which I'll deal with later.  For now, we can move on to sorting and subsetting.

## 2. Subsetting and Sorting

Pandas makes it easy to sort and subset dataframes because it is built on Numpy.  Since my dataframe is a bit cumbersome, let's cover subsetting first.  Subsetting is the process of simplifying the shape of a dataframe.

In [133]:
# To subset a dataframe's columns, just pass in the columns of interest, like this
MyFeatures = AllData[['date', 'HRV4T_Recovery_Points', 'training', 'trainingRPE', ' trainingTSS', 'alcohol','sleep_time']]

# To subset a dataframe's rows, specify expressions on columns of interest, like this
MyFeat_MyRows = MyFeatures[(MyFeatures['date'] >= '2020-01-01') & (MyFeatures['date'] <= '2020-05-31') & (MyFeatures['training'].isin(['easy','average','intense']))]


The subsetted dataframe MyFeat_MyRows will now show the seven columns of interest, and only my easy, average and intense training days between January 1 and May 31, 2020.  Subsetting and slicing dataframes is made even easier by using explicit indexing, and we'll cover that in section 4 below.  For now, let's look at sorting the dataframe.

In [134]:
# To sort by a single column
MyFeat_MyRows.sort_values('date', ascending = True)

# To sort by multiple columns, and change the order in which they are sorted, pass in lists like this
MyFeat_MyRows.sort_values([' trainingTSS', 'trainingRPE'], ascending = [False, True])

date  HRV4T_Recovery_Points training  trainingRPE   trainingTSS  \
176  2020-05-17                   6.95  intense          9.7         122.0   
169  2020-05-10                   6.90  intense          9.2         118.0   
162  2020-05-03                   7.00  intense          8.4         114.0   
120  2020-03-22                   7.26  intense          9.3         113.0   
113  2020-03-15                   7.04  average          8.6         111.0   
..          ...                    ...      ...          ...           ...   
86   2020-02-17                   7.13  intense          9.2          34.0   
184  2020-05-25                   7.05  intense          9.3          34.0   
163  2020-05-04                   6.74     easy          3.0          33.0   
150  2020-04-21                   7.18  intense          9.5          31.0   
66   2020-01-05                    NaN  average          3.0           NaN   

      alcohol  sleep_time  
176  a little         9.0  
169   nothing         8.7  
162  a little         8.2  
120  a little         8.9  
113   nothing         8.2  
..        ...         ...  
86    nothing         8.2  
184  a little         6.8  
163   nothing         7.9  
150  a little         7.2  
66    nothing         0.0  

[78 rows x 7 columns]

In [135]:
# To sort by a single column
MyFeat_MyRows.sort_values('date', ascending = True)

# To sort by multiple columns, and change the order in which they are sorted, pass in lists like this
MyFeat_MyRows.sort_values([' trainingTSS', 'trainingRPE'], ascending = [False, True])

date  HRV4T_Recovery_Points training  trainingRPE   trainingTSS  \
176  2020-05-17                   6.95  intense          9.7         122.0   
169  2020-05-10                   6.90  intense          9.2         118.0   
162  2020-05-03                   7.00  intense          8.4         114.0   
120  2020-03-22                   7.26  intense          9.3         113.0   
113  2020-03-15                   7.04  average          8.6         111.0   
..          ...                    ...      ...          ...           ...   
86   2020-02-17                   7.13  intense          9.2          34.0   
184  2020-05-25                   7.05  intense          9.3          34.0   
163  2020-05-04                   6.74     easy          3.0          33.0   
150  2020-04-21                   7.18  intense          9.5          31.0   
66   2020-01-05                    NaN  average          3.0           NaN   

      alcohol  sleep_time  
176  a little         9.0  
169   nothing         8.7  
162  a little         8.2  
120  a little         8.9  
113   nothing         8.2  
..        ...         ...  
86    nothing         8.2  
184  a little         6.8  
163   nothing         7.9  
150  a little         7.2  
66    nothing         0.0  

[78 rows x 7 columns]

Great.  I've now sorted my subsetted dataframe in descending order by 'trainingTSS' (which is a measure of how stressful my training was for that day, both in terms of intensity and duration) and ascending order by 'trainingRPE' (which is a subjective measure of my real perceived exertion of the training, on a scale of 1 to 10).  I can see that some of my most stressful training days were perceived as such, but many were not.  Hmmmm.  Let's look at some statistics.

## 3. Calculating Summary Statistics

We typically want to calculate summary (or descriptive) statistics for various measures along various dimensions in our dataframe.  For example, to calculate my mean trainingRPE for each type of training day (training), I could use

In [136]:
MyFeat_MyRows.groupby('training')['trainingRPE'].mean()

training
average    6.095238
easy       3.151429
intense    8.531818
Name: trainingRPE, dtype: float64

In addition to .mean(), there are a variety of other central tendency (.median(), .mode()), and dispersion (.min(), .max(), .var() and .std()) statistics we can use in Pandas.  We can also compute statistics for many measures and dimensions at once, by passing lists.  Here, I'll calculate the mean of 4 measures across the values of 2 dimensions, training and alcohol (my beer consumption the evening before a training day, where 'a little' = 1 beer, and 'too much' = 2 beers).   

In [137]:
MyFeat_MyRows.groupby(['training', 'alcohol'])[['HRV4T_Recovery_Points', 'trainingRPE', ' trainingTSS', 'sleep_time' ]].mean()

HRV4T_Recovery_Points  trainingRPE   trainingTSS  \
training alcohol                                                      
average  a little               6.962857     5.657143     59.285714   
         nothing                7.003636     6.338462     74.750000   
         too much               7.090000     6.000000     69.000000   
easy     a little               6.947273     3.000000     41.250000   
         nothing                6.922778     3.288889     44.555556   
         too much               6.820000     3.020000     43.200000   
intense  a little               7.038889     8.700000     81.555556   
         nothing                6.994167     8.415385     85.538462   

                   sleep_time  
training alcohol               
average  a little    7.657143  
         nothing     6.715385  
         too much    7.900000  
easy     a little    7.533333  
         nothing     8.027778  
         too much    6.520000  
intense  a little    7.833333  
         nothing     8.100000

## 4. Creating Pivot Tables

Pivot tables (using the .pivot_table method) allow us to compute many of the same descriptive statistics as the .groupby method above, but for one measure at a time.  For example, to get the mean (default statistic) trainingRPE for each type of training as we did above, we pass 'trainingRPE' to the values argument, and 'training' to the index argument  

In [138]:
MyFeat_MyRows.pivot_table(values = 'trainingRPE', index = 'training')

trainingRPE
training             
average      6.095238
easy         3.151429
intense      8.531818

Or we can calculate the mean training RPE for each type of training AND alcohol consumption category, by using the columns argument as well.  Note:  Since my dataset contains null values for some instances of alcohol use, I can use fill_value argument to specify how to treat these.  Here, I'll just return blanks.  And I can also request row and column-wise statistics using the margins argument.  

In [139]:
MyFeat_MyRows.pivot_table(values = 'trainingRPE', index = 'training', columns = 'alcohol', fill_value = '', margins = True)

alcohol  a little  nothing too much      All
training                                    
average   5.65714  6.33846        6  6.09524
easy            3  3.28889     3.02  3.15143
intense       8.7  8.41538           8.53182
All       5.49643  5.70455  3.51667  5.46154

By default, the .pivot_table method aggregates by mean, but we can compute different statistics, and even multiple statistics by passing the aggfunc argument.  Here, we can calculate the mean and median training RPE for each row and column dimension, as follows

In [140]:
MyFeat_MyRows.pivot_table(values = 'trainingRPE', index = 'training', columns = 'alcohol', aggfunc = [np.mean,np.median], fill_value = '', margins = True)

mean                              median                       
alcohol  a little  nothing too much      All a little nothing too much   All
training                                                                    
average   5.65714  6.33846        6  6.09524      6.6     6.5        6   6.5
easy            3  3.28889     3.02  3.15143        3       3        3     3
intense       8.7  8.41538           8.53182      9.3     8.6            8.6
All       5.49643  5.70455  3.51667  5.46154      4.4     5.4        3  4.85

## 5. Using Explicit Indexes

Remember how we subsetted our original dataframe AllData into our columns and rows of interest above in section 2?  We can accomplish many of the same objectives more efficiently by using indexing (using .loc).  Slicing and subsetting using .loc is extremely powerful, but first requires you to set one or more columns in your dataframe as your index, and then sort the dataframe using those indexes.  For example, to subset easy, average or intense training days from our original 7-column dataframe (MyFeatures) we can set 'training' as our index, and use .loc to subset our training days of interest 

In [141]:
# Set index to our column of interest
MyFeatures_ind = MyFeatures.set_index ('training')

# Subset our frame on the 'training' days of interest
MyFeatures_subsetted = MyFeatures_ind.loc[['easy','average','intense']]

In our new dataframe MyFeatures_subsetted, we could then subset on our dates of interest (January through May, 2020) by setting 'date' as our index, and using .loc to get our dates

In [142]:
# First, reset our 'training' column index
MyFeatures_subsetted.reset_index('training', inplace=True)

# Then set our new index based on 'date', and sort by the index
MyFeatures_subsetted_ind = MyFeatures_subsetted.set_index ('date').sort_index()

# Subset our frame by the dates of interest
MyFeatures_subsetted_ind.loc['2020-01-01':'2020-05-31']

training  HRV4T_Recovery_Points  trainingRPE   trainingTSS  \
date                                                                    
2020-01-05  average                    NaN          3.0           NaN   
2020-01-06  intense                    NaN          8.6          70.0   
2020-01-31  average                    NaN          4.2          54.0   
2020-02-03     easy                   7.13          3.0          39.0   
2020-02-04     easy                   6.85          3.0          39.0   
...             ...                    ...          ...           ...   
2020-05-24     easy                   6.80          5.2          73.0   
2020-05-25  intense                   7.05          9.3          34.0   
2020-05-26     easy                   7.00          3.0          39.0   
2020-05-28  intense                   7.27          7.9          91.0   
2020-05-31  intense                   7.09          7.3         103.0   

             alcohol  sleep_time  
date                              
2020-01-05   nothing         0.0  
2020-01-06   nothing         8.0  
2020-01-31   nothing         0.0  
2020-02-03   nothing         7.5  
2020-02-04  a little         8.2  
...              ...         ...  
2020-05-24   nothing         8.6  
2020-05-25  a little         6.8  
2020-05-26  a little         9.7  
2020-05-28   nothing         6.5  
2020-05-31   nothing         8.7  

[78 rows x 6 columns]